### ISSUE TITLE
Bootstrap tabs WAI-ARIA attributes set for better accessibility

### ISSUE THREAD

**Issue by bassettsj** [twbs/bootstrap#13554]
Again from the [paypal/bootstrap-accessibility-plugin](https://github.com/paypal/bootstrap-accessibility-plugin/#tab-panel), these changes should be added to the documentation and the JS plugin.

### Tab Panel

1. ~~Add ARIA roles like tablist, presentation, and tab for tabs UL, LI.~~
1. Add tabIndex, ~~aria-expanded~~, aria-selected, ~~aria-controls~~ for tab.
1. Add ARIA roles of ~~tabpanel~~, tabIndex, aria-hidden, and aria-labelledBy for tabPanel.
1. Add keydown event listener for the tab to work with keyboard.
1. Dynamically flip tabIndex, aria-selected, and ~~aria-expanded~~ for tab when it is activated and add aria-hidden to hide the previously visible tab.

**Comment by goetsu** [comment]
the role="tabpanel" doesn't need to be applied to the parent div it's only needed for the content div  : 
```
<div role="tabpanel">
<ul class="nav nav-tabs" role="tablist">...</ul>
<div class="tab-content">
<div id="overview" class="tab-pane overview" role="tabpanel">...</div>
...
</div>
</div>
```
must be : 
```
<div>
<ul class="nav nav-tabs" role="tablist">...</ul>
<div class="tab-content">
<div id="overview" class="tab-pane overview">...</div>
...
</div>
</div>
```

**Comment by patrickhlauke** [comment]
@goetsu i believe that is incorrect. please refer to the examples for http://www.w3.org/TR/wai-aria-practices/#tabpanel - e.g. http://www.oaa-accessibility.org/examplep/tabpanel1/

**Comment by goetsu** [comment]
that's what i do and there is no role tabpanel on the parent div <div id="tabpanel1" class="tabpanel">

**Comment by patrickhlauke** [comment]
and there isn't a ``role="tabpanel`` on the parent div in http://getbootstrap.com/javascript/#markup either, so not sure what you're driving at here.

**Comment by goetsu** [comment]
<div role="tabpanel">

**Comment by goetsu** [comment]
```
<div role="tabpanel">

  <!-- Nav tabs -->
```

**Comment by patrickhlauke** [comment]
Ah, now I see what you mean. Your second code example removes ``role="tabpanel"`` from the actual tab panel itself, so that's quite confusing - and that part is clearly wrong in your proposed change.

Can I suggest that you don't simply append comments to issues like this in future, and please file an appropriate separate issue, being quite specific about what you actually mean?

**Comment by goetsu** [comment]
ok sorry for that and I agree that the actual tab panel itself need role="tabpanel"

**Comment by mdo** [comment]
What's the status of this one? Does a lot of stuff still need addressing? What can be done in v3 vs v4?

/cc @patrickhlauke

**Comment by patrickhlauke** [comment]
Need some time to review what our current state of tabs is, as I've admittedly lost track. I'm guessing if we're still missing something related to attributes, that'll be fixed with my plan to automagically inject them.

**Comment by peterh-capella** [comment]
I've been doing some work recently with adding additional accessibility and flexibility to the already good work done with tabs. One thing relevant here is changing the attribute on the tab from `aria-expanded` to `aria-selected` to conform to [standards](http://www.w3.org/TR/wai-aria/roles#tab). ("In either case, authors SHOULD ensure that a selected tab has its aria-selected attribute set to true....") I also abstracted the markup a bit to allow tabs that are not structured as a list. (Recently found that some screen-reader users prefer this, particularly for site navigation. See [this CSS-tricks post](https://css-tricks.com/navigation-in-lists-to-be-or-not-to-be/).) Other work I'm doing looks like it addresses some of the issues in the original post.

So this is basically following Step 1 of the Pull Request contribution guidelines: asking first if my work and contribution would be useful.

**Comment by mdo** [comment]
[**Bootstrap 3 is no longer being officially developed or supported.**](https://github.com/twbs/bootstrap/issues/20631)

All work has moved onto our next major release, v4. As such, this issue or pull request is being closed as a "won't fix." For additional help and support, we recommend [utilizing our community resources](https://github.com/twbs/bootstrap/blob/master/README.md#community). Thanks for your understanding, and see you on the other side of v4!

<3,
@mdo and team
